auth: Support parallel fingerprint auth

moggiesir opened the pull request:
I chose to use Fprint's dbus interface directly rather than going through pam (which uses Fprint's dbus interface) due to poor handling of system sleep somewhere between fprintd and pam. When preparing for sleep, fprintd puts the device to sleep, which causes VerifyStatus to emit with verify-unknown-error, which normally should be responded to by calling both Device.StopVerify and Device.Release (and this is what pam does). Unfortunately, if you try to release the device when the system is preparing for sleep, you'll get an error that the device is busy and then you can't can't claim or release the device for 30 seconds.

pam also has a max timeout for pam_fprintd.so of 99 seconds, and so if we used pam, we'd have to deal with the timeouts and keep restarting the auth conversation.

gdm/gnome-session lock seems to get around these issues by having a shutter on top of the lock screen that you have to interact with first that gives gnome-session a trigger to start fingerprint auth.

Fixes #258

PaideiaDilemma commented on src/core/Fingerprint.cpp:
```diff
@@ -0,0 +1,237 @@
+                m_sDBUSState.retries++;
+                m_sDBUSState.failText = RETRY_MESSAGE;
+                if (clearRetryTextTimer) clearRetryTextTimer->cancel();
+                clearRetryTextTimer   = g_pHyprlock->addTimer(
+                    std::chrono::seconds(2),
```
This could just be the `input-field:fail_timeout` option I think?

moggiesir replied:
It's not straightforward AFAICT since input-field is a special category according to the config so I'm not sure how to get it aside from looping through the widget configs and finding the input-field, if it's there, but let me know if I'm missing something.

PaideiaDilemma replied:
You are right.

Idea: Add `$FPRINTPROMPT` and `$FPRINTFAIL` like `$PROMPT` and `$FAIL` in `Auth.cpp` (`getLastPrompt` and `getLastFailText`), then we would not need the timer at all. Although you would not be able to have `$FPRINTFEEDBACK` like behavior. Currently you also don't have that for normal auth. Switching between fail and prompt is input-field exclusive.

moggiesir replied:
There's a signal for finger presence, so I switched to change the $FPRINTMESSAGE at that time, removing the need for the timer.

PaideiaDilemma commented on src/core/Fingerprint.cpp:
```diff
@@ -0,0 +1,237 @@
+    return m_sDBUSState.failText.empty() ? std::nullopt : std::optional(m_sDBUSState.failText);
+}
+
+void CFingerprint::terminate() {
+    m_sDBUSState.connection->leaveEventLoop();
+}
```
need a check here?

PaideiaDilemma replied:
`m_sDBUSState.connection` is a nullptr when fingerprint is not enabled right?

moggiesir replied:
Done

Kochanac commented:
Hello, tested this feature, here is what I think:

- when trying to use the wrong finger 3 times, hyprlock exits and unlocks the session (logs ([hyprlock -v](https://github.com/user-attachments/files/17355829/hyprlock_logs.txt)) and my config [hyprlock.conf.txt](https://github.com/user-attachments/files/17355837/hyprlock.conf.txt), compiled from a2ef6a9907f6c26720150af730cc81fed7f82688)

- maybe this is my config problem, but anyway: I don't see any error messages when getting verify-no-match error

- from a UX standpoint I would suggest that as an option you could add some message when starting to verify using fprint, so a user could see that the system is working correctly

moggiesir commented on src/core/Fingerprint.cpp:
```diff
@@ -0,0 +1,237 @@
+#include <security/pam_appl.h>
+#if __has_include(<security/pam_misc.h>)
+#include <security/pam_misc.h>
+#endif
```
Removed

moggiesir commented:
> Hello, tested this feature, here is what I think:
> 
> * when trying to use the wrong finger 3 times, hyprlock exits and unlocks the session (logs ([hyprlock -v](https://github.com/user-attachments/files/17355829/hyprlock_logs.txt)) and my config [hyprlock.conf.txt](https://github.com/user-attachments/files/17355837/hyprlock.conf.txt), compiled from [a2ef6a9](https://github.com/hyprwm/hyprlock/commit/a2ef6a9907f6c26720150af730cc81fed7f82688))

This is now fixed, it was a bad state during some refactoring.

> * probably this is my config problem, but anyway: I don't see any error messages in the input field when getting verify-no-match error

I didn't want to mess with the input-field so as to not introduce edge cases with the password input/error handling, so you can use $FPRINTMESSAGE in a label.

> * from a UX standpoint I would suggest that as an option you could add some message when starting to verify using fprint, so a user could see that the system is working correctly

I added an option general:fingerprint_ready_message that $FPRINTMESSAGE will display when ready for fingerprint input.

Kochanac commented:
Indeed the bug is fixed, thanks!

About fingerprint_ready_message, I meant another thing, sorry for the unclear explanation. I've meant that when you touch the scanner, the system should immediately display some message like "scanning..." to indicate that scan is started. I think it would be useful for improving responsiveness as sometimes the scan can be longer than 500ms

But it looks like fprint does not expose such interface, so it is probably impractical to do. fingerprint_ready_message is also useful, so I would leave it as is

PaideiaDilemma commented:
> I didn't want to mess with the input-field so as to not introduce edge cases with the password input/error handling, so you can use $FPRINTMESSAGE in a label.

In my opinion it would make sens to be able to have the fp feedback in the input-field, if it is empty. But for that we might want to add an abstraction so that we don't have to add logic for an additional auth module in the input-field. That could be done later.

moggiesir commented:
> About fingerprint_ready_message, I meant another thing, sorry for the unclear explanation. I've meant that when you touch the scanner, the system should immediately display some message like "scanning..." to indicate that scan is started. I think it would be useful for improving responsiveness as sometimes the scan can be longer than 500ms
> 
> But it looks like fprint does not expose such interface, so it is probably impractical to do. fingerprint_ready_message is also useful, so I would leave it as is

There's a property on the device for whether the finger is present, so I changed to use the PropertiesChanged signal to listen for finger presence and display a message at that time (general:fingerprint_present_message). On my device, it's pretty snappy to read my fingerprint, so I don't like the extra messages.

PaideiaDilemma commented on src/core/Fingerprint.cpp:
```diff
@@ -0,0 +1,240 @@
+#include "Fingerprint.hpp"
+#include "../helpers/Log.hpp"
+#include "src/config/ConfigManager.hpp"
```
Relative include pls

moggiesir replied:
Done

PaideiaDilemma commented on src/core/Fingerprint.cpp:
```diff
@@ -0,0 +1,240 @@
+}
+
+static void fingerprintUnlockCallback(std::shared_ptr<CTimer> self, void* data) {
+    g_pHyprlock->unlock();
```
I don't really like that, because i think the check for `m_bAuthenticated` should be right here. Just make it a `onFingerprintCheckTimer` like with the password authentication and call it whether or not we are authenticated.

moggiesir replied:
Done

PaideiaDilemma commented on src/core/Fingerprint.cpp:
```diff
@@ -0,0 +1,240 @@
+
+    registerSleepHandler();
+
+    if (m_sDBUSState.sleeping) {
+        m_sDBUSState.connection->enterEventLoop();
+    } else if (startVerify()) {
+        m_sDBUSState.message = m_sFingerprintReady;
+        g_pHyprlock->enqueueForceUpdateTimers();
+        m_sDBUSState.connection->enterEventLoop();
```
NITPICK: would be a bit easier to understand if `enterEventLoop` would not happen in `setupDBUS`, but either in the std::thread ctor or in a dedicated `enterLoop` function.

moggiesir replied:
Cleaned this up a bit and moved the logic into the thread ctor.

PaideiaDilemma commented on src/core/Fingerprint.cpp:
```diff
@@ -0,0 +1,240 @@
+        } else {
+            m_sDBUSState.sleeping = false;
+            inhibitSleep();
+            std::this_thread::sleep_for(std::chrono::seconds(1));
```
Can you add a comment for why this sleep is needed?

PaideiaDilemma commented on src/core/Fingerprint.cpp:
```diff
@@ -0,0 +1,240 @@
+        inhibitSleep();
+}
+
+void CFingerprint::inhibitSleep() {
+    m_sDBUSState.login->callMethod("Inhibit")
+        .onInterface(LOGIN_MANAGER)
+        .withArguments("sleep", "hyprlock", "Release fingerprint before sleep", "delay")
+        .storeResultsTo(m_sDBUSState.inhibitLock);
+}
```
So this is used to make sure we call `stopVerify` before we are going to sleep right?

moggiesir replied:
Right, updated the inhibit reason.

PaideiaDilemma commented on src/core/Fingerprint.cpp:
```diff
@@ -0,0 +1,240 @@
+void CFingerprint::handleVerifyStatus(const std::string& result, bool done) {
+    Debug::log(LOG, "Status: {} {}", done, result);
+    auto matchResult = s_mapStringToTestType[result];
+    if (m_sDBUSState.sleeping && matchResult != MatchResult::Disconnected) {
```
Brackets can be omitted.

moggiesir replied:
Done

PaideiaDilemma commented on src/core/Fingerprint.cpp:
```diff
@@ -0,0 +1,240 @@
+            stopVerify();
+            m_bAuthenticated     = true;
+            m_sDBUSState.message = "";
+            Debug::log(LOG, "matched");
```
This log and others could be improved. Ideally prefixed with `fprint: ` or smth.

moggiesir replied:
Done

PaideiaDilemma commented on src/core/Fingerprint.cpp:
```diff
@@ -0,0 +1,240 @@
+        case MatchResult::NoMatch:
+            stopVerify();
+            if (m_sDBUSState.retries < 2) {
+                startVerify();
+                done = false;
+                m_sDBUSState.retries++;
+                m_sDBUSState.message = RETRY_MESSAGE;
```
Check whether `startVerify` succeeded? otherwise that would cause a silent fail I think.

moggiesir replied:
Done

PaideiaDilemma commented:
The rest looks pretty good.
Thanks for working on this! Sadly I cannot properly test it, because my fp reader has no fprintd driver ;(

noxxious commented:
The PR needs changes to flake.nix to add extra buildInputs: sdbus-cpp and systemdLibs.

PaideiaDilemma commented on src/core/Fingerprint.cpp:
```diff
@@ -0,0 +1,232 @@
+    if (!m_bEnabled)
+        return;
+    std::thread([this]() {
+        // For grace or SIGUSR1 unlocks
+        if (g_pHyprlock->isUnlocked())
+            return;
```
That does not hurt but I think it is also not needed.

moggiesir replied:
Removed

b5l replied:
> That does not hurt
> **I think** it is also not needed

Unless you **know**, I consider this a reason for keeping those lines in.

PaideiaDilemma replied:
well i know that **should not** be needed, because a race, where we go through that check and then hyprlock is unlocked, (via grace or USR1) might cause some problems. So even if the check fixes smth, it does not fix it properly.

PaideiaDilemma replied:
@moggiesir is it fine when the main thread calls `terminate` before `m_sDBUSState.connection = sdbus::createSystemBusConnection();` succeeds? Also is `leaveEventLoop` thread safe?

moggiesir replied:
It's fine when the main thread calls `terminate` before `m_sDBUSState.connection = sdbus::createSystemBusConnection();` succeeds since `terminate` checks if the pointer is initialized. I also made sure it's fine to call `leaveEventLoop` even if `enterEventLoop` was never called.

For `leaveEventLoop` thread safety, from the [docs](https://github.com/Kistler-Group/sdbus-cpp/blob/master/docs/using-sdbus-c++.md):

> An event loop that blocks in the synchronous enterEventLoop() call can be unblocked through leaveEventLoop() call on the respective bus connection issued from a different thread or from an OS signal handler.

Further, I changed to return the thread from `CFingerprint::start()` so that hyprlock will wait for the thread to terminate to make sure we don't try to destroy the dbus connection while some call is being made.

PaideiaDilemma replied:
Nice, thanks!

matejdro commented:
I have did a quick check of this using fingerprint sensor in the ThinkPad T14s (AMD) Gen 3 and everything seem to be working well. Thanks

PaideiaDilemma commented on src/config/ConfigManager.cpp:
```diff
@@ -50,6 +50,9 @@ void CConfigManager::init() {
     m_config.addConfigValue("general:pam_module", Hyprlang::STRING{"hyprlock"});
     m_config.addConfigValue("general:fractional_scaling", Hyprlang::INT{2});
+    m_config.addConfigValue("general:enable_fingerprint", Hyprlang::INT{0});
+    m_config.addConfigValue("general:fingerprint_ready_message", Hyprlang::STRING{""});
+    m_config.addConfigValue("general:fingerprint_present_message", Hyprlang::STRING{""});
```
add sensible defaults for the messages?

moggiesir replied:
Done

PaideiaDilemma commented:
@vaxerski I think this looks good

vaxerski commented on src/core/Fingerprint.cpp:
```diff
@@ -0,0 +1,228 @@
+std::optional<std::thread> CFingerprint::start() {
+    if (!m_bEnabled)
+        return {};
+    return std::thread([this]() {
```
absolutely do _not_ multithread here. Get the pollfd from sdbus-cpp and integrate it within the main event loop (see xdph: https://github.com/hyprwm/xdg-desktop-portal-hyprland/blob/fb9c8d665af0588bb087f97d0f673ddf0d501787/src/core/PortalManager.cpp#L270)

moggiesir replied:
Done. I was originally trying to maintain more consistency with CAuth.

It makes things a bit more complicated, but I kept it so that the dbus connection is only initiated/listened to if fingerprint auth is enabled. I'd be happy to change it to always create the connection if you'd rather that.

vaxerski commented on CMakeLists.txt:
```diff
@@ -56,7 +56,8 @@ pkg_check_modules(
   gbm
-  hyprutils>=0.2.3)
+  hyprutils>=0.2.3
+  sdbus-c++)
```
you are using sdbus-c++ 2 here, so please add `>=2.0.0` to not confuse packagers (xdph still uses 1.x)

moggiesir replied:
What am I using from sdbus-c++>=2.0.0? On my system, I only have 1.5.0 installed.

vaxerski replied:
oh. I thought the syntax looked 2.0-like. Use 2.0 then. If you have a system 1.5, either install 2.0 or add a subproject and check it out to the latest release.

2.0 is breaking and I don't want to "fix it later", already lazy to do it with xdph

moggiesir replied:
Done.

vaxerski commented on src/core/Fingerprint.hpp:
```diff
@@ -0,0 +1,70 @@
+    void                                terminate();
+
+  private:
+    enum class MatchResult {
```
C++: `enum`
Also, please:
- call enum variables in CAPS and with an identifiable prefix (e.g. `MATCH_NO_MATCH`, `MATCH_RETRY`)
- add a variable for invalid e.g. `MATCH_INVALID = 0`
- have the _first_ enum value be defined (e.g. as above = 0)

moggiesir replied:
Done

vaxerski commented on src/core/Fingerprint.hpp:
```diff
@@ -0,0 +1,70 @@
+    bool                               m_bAuthenticated = false;
+    bool                               m_bEnabled       = false;
+
+    std::map<std::string, MatchResult> s_mapStringToTestType = {{"verify-no-match", MatchResult::NoMatch},
```
can this be moved to .cpp and static'd?

moggiesir replied:
Done

vaxerski commented on src/core/hyprlock.cpp:
```diff
@@ -425,16 +430,21 @@ void CHyprlock::run() {
+        .fd     = wl_display_get_fd(m_sWaylandState.display),
+        .events = POLLIN,
     };
+    if (conn)
```
{}, this is not a short if

moggiesir replied:
Done

vaxerski reviewed (commented):
minor nitpicks, rest is alr, needs a wiki mr

moggiesir commented:
Opened https://github.com/hyprwm/hyprland-wiki/pull/816 for the wiki

moggiesir commented:
Yes, with the new parallel auth support, users should go back to the default pam file.

vaxerski commented on src/core/Fingerprint.hpp:
```diff
@@ -0,0 +1,52 @@
+#include <memory>
+#include <optional>
+#include <string>
+#include <condition_variable>
```
include not needed anymore?

moggiesir replied:
Done

vaxerski commented:
apart from that lgtm

PaideiaDilemma commented:
@vaxerski

Quite a few fixes since last release. I think a hyprlock release is due.
Wanna do that before or after this?

vaxerski commented:
I need @fufexan to fix nix ci and then we can release

vaxerski commented:
man I love fufexan
